tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,769 @@
+package tensorflow
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"reflect"
+	"runtime"
+	"unsafe"
+
+	"github.com/golang/protobuf/proto"
+
+	pb "github.com/tensorflow/tensorflow/tensorflow/contrib/go/proto"
+)
+
+import "C"
+
+const (
+	cAckByte  = 6
+	cBellByte = 7
+	cDc1      = 17
+
+	cBytesComplex64 = 8
+	cBytesFloat32   = 4
+	cBytesFloat64   = 8
+	cBytesInt16     = 2
+	cBytesInt32     = 4
+	cBytesInt64     = 8
+	cBytesUint16    = 2
+)
+
+// A DataType represents the type of the data contained in a Tensor
+type DataType pb.DataType
+
+// A TensorInterface is the interface implemented by Tensors.
+type TensorInterface interface {
+	Data() []byte
+	DataSize() int64
+	DataType() DataType
+	GetVal(d ...int) (val interface{}, err error)
+
+	Dim(n int) int
+	NumDims() int
+
+	Bool() (res []bool, err error)
+	Float32() (res []float32, err error)
+	Float64() (res []float64, err error)
+	Int32() (res []int32, err error)
+	Int64() (res []int64, err error)
+	Str() (res [][]byte, err error)
```

Comment:
I think that the 's' at the end can help to understand that is more than one and that the returned value is a slice, so my vote for the 's' and ByteSlices as exception.
Regarding to the String method, it is just a helper used to print the Tensor in a human readable way:
> If an operand implements method String() string, that method will be invoked to convert the object to a string, which will then be formatted as required by the verb (if any)
https://golang.org/pkg/fmt/